### DISCUSSION TITLE
Question to recent change `Records<K, V>` to `{ [key: K]: V }`

### DISCUSSION THREAD

**Discussion by Erisli** [Aleph-Alpha/ts-rs#337] [Q&A] [answered]
Recently, `Records<K, V>` was changed to `{ [key: K]: V }`. This works well with `HashMap<String, T>`, but not really well with `HashMap<Enum, T>`. When I imported this type to my ts, it gave me this error

`An index signature parameter type cannot be a literal type or generic type. Consider using a mapped object type instead.`

I tried changing it from

```ts
type MyType = { [key: MyEnum]: MyStruct } 
```

To

```ts
type MyType = { [key in MyEnum]: MyStruct } 
```

It works.

I have also tried doing  `#[ts(type = "...")]`, but I am having trouble of importing `MyStruct` and `MyEnum` automatically. 
 
 Is there any suggestion to go around this without manually implementing it? Thank you all for such great work.

**Comment by gustavo-shigueo** [comment, marked as answer]
Hello @Erisli!

This change was made to allow recursive types when dealing with `HashMap`. I'll quickly check if changing to a mapped type breaks anything and if not, I'll open a PR

**Reply by gustavo-shigueo** [reply]
@Erisli can you provide the definitions of `MyEnum` and `MyStruct` so that I can create a test case for the issue?

**Reply by Erisli** [reply]
Certainly!

```rs
pub struct MyType {
    pub MyField_0: bool,
    pub MyField_1: HashMap<MyEnum, MyStruct>,
}
```

```rs
pub enum MyEnum {
    Variant_0,
    Variant_1,
    Variant_2,
    Variant_3,
}
```

```rs
pub struct MyStruct {
    pub MyField_0: bool,
    pub MyField_1: u32,
    pub MyField_2: Option<u32>,
    pub MyField_3: Option<u32>,
    pub MyField_4: Option<u32>,
    pub MyField_5: String,
}
```

**Reply by gustavo-shigueo** [reply]
Thanks!

**Reply by gustavo-shigueo** [reply]
I had to change it so the type will generate as
```ts
type MyType = { [key in MyEnum]?: MyStruct }
//                             ^ Note the `?`
```

Otherwise TS would require ALL variants to be in the object

**Reply by Erisli** [reply]
Looks good to me!

**Reply by gustavo-shigueo** [reply]
Done! Check out #339

**Reply by Erisli** [reply]
Thank you again for such amazing work and project!!!!!
